### PR TITLE
No-Deploy-Friday-Nudge #56

### PR DESCRIPTION
**Pull request by xrvdg**
This pull request adds (non-configurable) No-Deploy-Friday-Nudge functionality (#56) to Hoff.

### CODE REVIEW

File: src/Logic.hs
Comment (ReinierMaas) [inline src/Logic.hs:444]: This should be moved to `Project.hs`

File: src/Logic.hs
Comment (ReinierMaas) [inline src/Logic.hs:453]: Maybe a clearer name is `handleMergeRequested`?

File: src/Logic.hs
Comment (ReinierMaas) [inline src/Logic.hs:566]: There is a function `needsDeploy` in `Project.hs` that provides this information.

File: src/Project.hs
Comment (ReinierMaas) [inline src/Project.hs:413]: We often choose between two different indentation styles. Having the `=` symbol aligned vertically (on two spaces) or having a single space after the code:

```Haskell
fn Match     = undefined
fn MatchLong = undefined
```

```Haskell
fn Match = undefined
fn MatchLong = undefined
```

Can you choose one or the other?

File: tests/Spec.hs
Comment (ReinierMaas) [inline tests/Spec.hs:110]: Can you make this a list so we can give concurrent calls to the function increasing timestamps?

File: tests/Spec.hs
Comment (ReinierMaas) [inline tests/Spec.hs:125]: Can you default to repeat the given timestamp?

File: tests/Spec.hs
Comment (ReinierMaas) [inline tests/Spec.hs:219]: This would then need to be updated to `takeResultGetDateTime`.

File: src/Logic.hs
Comment (ReinierMaas) [reply src/Logic.hs:433]: Should we allow `merge on Friday` on other days?

**Review by ReinierMaas — CHANGES_REQUESTED** [review]
The overall design looks good to me. I just have style pointers but nothing really major. Using the free monad to encapsulate getting the timestamps nicely abstracts it for the testing! And the additional tests also look extensive.

File: src/Logic.hs
Comment (fatho) [inline src/Logic.hs:388]: A minor thing: I wonder if "on" is a good choice of words here, as it could be mistaken for a scheduled merge (which we don't support). Maybe "even on"? Not really sure, would otherwise also be fine.

File: src/Project.hs
Comment (fatho) [inline src/Project.hs:109]: Design-wise, I'm not sure this is the right way to go, as it results in a combinatorial blow-up when more options are added.

The fact that it's Friday doesn't change anything inherently about the command, a `MergeAndDeploy` will still merge and deploy. What changes is the way how that command is parsed. I think something like our original `ApprovedFor` together with a `Fridayness` type (e.g. `data Fridayness = It'sFriday | It'sNotFriday`) in a `(ApprovedFor, Fridayness)` tuple would reflect more accurately what is going on and avoid the blow-up.

**Review by fatho — COMMENTED** [review]
Your first PR :tada: Besides Reinier's remarks, I think the design can also still be simplified/future-proofed (see below).

File: src/Logic.hs
Comment (ruuda) [inline src/Logic.hs:425]: This gets the date in the UTC time zone. It will be Friday two hours later in Europe/Amsterdam than in UTC when DST is in effect. Is that intentional? (And also, does it make sense to pick midnight as the boundary? Is Saturday 04:00 an okay time to merge things?)

Comment (ReinierMaas) [reply src/Logic.hs:426]: Thanks for the feedback. What would be your proposal for updating the nudge? Should we detect whether it is Friday [Anywhere on Earth](https://en.wikipedia.org/wiki/Anywhere_on_Earth)? (Start at UTC+14 / End at UTC-12)
Using AoE as the dedicated time range would already disallow the merge to proceed on Thursday afternoon in Europe/Amsterdam when DST is in effect. Which might be a valid moment for the stop merging nudge.

Comment (fatho) [reply src/Logic.hs:426]: Good point about the timezone! I think just detecting Friday at our location is good enough (maybe the timezone could be made configurable, but that's about it).

> Is Saturday 04:00 an okay time to merge things

It is not. But the feature is intended as a guardrail against accidents, not a foolproof method of preventing all untimely merges, and I'd just assume that there won't be accidental merges on Saturday at 04:00.

We might want to just cover the whole weekend, but I think nighttime can be ignored for the moment.

File: src/Logic.hs
Comment (fatho) [inline src/Logic.hs:444]: `s/it not/it is not/`

File: src/Logic.hs
Comment (fatho) [inline src/Logic.hs:445]: Good idea rejecting this as well so that people don't get conditioned to just always using the more powerful command :+1:

File: src/Logic.hs
Comment (fatho) [inline src/Logic.hs:427]: Can you open a new hoff issue about potentially making this more configurable (e.g. the timezone and the range)?
Then we'll also immediately have a good first issue for the next person joining :-)

**Review by fatho — COMMENTED** [review]
Just some small things from my side:

File: src/Logic.hs
Comment (ruuda) [reply src/Logic.hs:428]: > What would be your proposal for updating the nudge?

If it’s just a nudge on Fridays for people working in Europe/Amsterdam, I would just ignore the issue and keep it like this.

At the expense of scope creep, maybe a nicer solution that also covers the weekend is to provide a time range in the configuration file. It could include a start day of week + time of day and an end day of week + time of day, and a time zone in which to interpret them.

File: src/Logic.hs
Comment (ReinierMaas) [inline src/Logic.hs:451]: Nit: this line is very long.

**Review by ReinierMaas — APPROVED** [review]
LGTM! After resolving Fabians comments and one nit. Thanks for implementing this!

**Comment by xrvdg** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @xrvdg, rebasing now.

**Comment by OpsBotPrime** [comment]
Rebased as 991e117f494cb643d1532219642b101226ad24fe, waiting for CI …

**Pull request merged**
